### ISSUE TITLE
[BUG] device synchronize for async allocator can be really bad for performance

### ISSUE THREAD

**Issue by revans2** [NVIDIA/spark-rapids#9130]
**Describe the bug**
I am not really sure how common this is, but I was running the scale tests. I generated data with a scale of 10 and a complexity of 500, and then ran q1, q2, and q3 (which are still WIP at https://github.com/NVIDIA/spark-rapids/pull/9089) As a side note I did boost the timeout to 10 mins.

I did this on a 6 core/12 thread CPU with an a6000 configured for 4 tasks in parallel, 8 GiB of pinned memory and a spill size of 32 GiB, although only a few GB would have been enough.

When I ran the tests the time to run the query would vary wildly. In some cases it would be 2.5 min and in other cases it would be 7.5 mins.  For q3 it even timed out a few times at 10 mins. The odd thing was that all of the time came from what appeared to be device synchronize as a part of the async allocator.  On slow runs I would see what appeared to be a tight loop of retrying allocations with a device synchronize when nothing was spillable. The GPU utilization, according to nvtop, was about 65% to 75%, but the power usage was only at about 100 watts. Also the CPU utilization for the java process was about 60% of a single core. The fast runs didn't have nearly as many device synchronizes and would have the device utilization spike to 100% then fall to 0% for a while, somewhat regularly. But the power usage showed about 180 to 300 watts.

The odd thing is that the spill was the same between the two runs and the spill time/spill read time, was also almost the same between the two runs. It is as if the device synchronize slowed down the processing in other threads massively. Not 100% sure about that. I will try to get an nsys trace of what is happening.

**Comment by revans2** [comment]
I got an nsys trace of it happening, and it looks like a join gather can do a lot of work without having the GPU semaphore. I saw all 12 threads doing things, many of them in join gathers without the semaphore held.

**Comment by revans2** [comment]
I should correct this a bit. It was not all 12 threads. Some of the threads were "unscheduled" but I did see some running without the GPU semaphore.

**Comment by revans2** [comment]
A little more information. The join itself was not the problem, although it showed up in the metrics as a problem because of the semaphore not being grabbed. Most of the time the join didn't even start to do anything. The problem appears to be the parquet writes. When 4 of them would go at once, we have enough memory to do the write, but only if we keep switching it between different threads. But the device synchronize is a very expensive way of doing this. I'm not 100% sure we can fix the problem without some help from the team doing the ASYNC allocator.

**Comment by abellina** [comment]
> But the device synchronize is a very expensive way of doing this. I'm not 100% sure we can fix the problem without some help from the team doing the ASYNC allocator.

Is the time to synchronize not accounted for in the `spillTime` metric? Or do you see that being high as well.

**Comment by revans2** [comment]
If I switch the target batch size from 2g to 1g it fixes the problem because it drops the memory pressure (took 100.279 seconds instead of 5+ mins)
If I switch the parallelism to 3 instead of 4 it also mostly fixes the problem (took 131.437 seconds)

Both together is even faster still 96.760 seconds.

Oddly if I drop the total memory allowed to 16 GiB instead of the full 48 available it is still faster to spill to host memory in this case than to continually do the device synchronize (247.655 seconds).

**Comment by revans2** [comment]
> Is the time to synchronize not accounted for in the `spillTime` metric? Or do you see that being high as well.

Why would it be a part of spillTime?  We don't grab the semaphore except when we are reading back in spilled data. I "fixed" the semaphore problem by just grabbing it in hasNext for the join iterator.

```
diff --git a/sql-plugin/src/main/scala/com/nvidia/spark/rapids/AbstractGpuJoinIterator.scala b/sql-plugin/src/main/scala/com/nvidia/spark/rapids/AbstractGpuJoinIterator.scala
index 76f6c1ed9..92003ac46 100644
--- a/sql-plugin/src/main/scala/com/nvidia/spark/rapids/AbstractGpuJoinIterator.scala
+++ b/sql-plugin/src/main/scala/com/nvidia/spark/rapids/AbstractGpuJoinIterator.scala
@@ -85,6 +85,7 @@ abstract class AbstractGpuJoinIterator(
     if (closed) {
       return false
     }
+    GpuSemaphore.acquireIfNecessary(TaskContext.get())
     var mayContinue = true
     while (nextCb.isEmpty && mayContinue) {
       if (gathererStore.exists(!_.isDone)) {
```

This is arguably not the right place to do this because we might grab it in some situations where we don't need to, but it is better than not grabbing it at the right time. I think there are likely a number of other places where we need to grab it too. I'll try to come up with a patch for it generally.

**Comment by abellina** [comment]
I was thinking about the synchronize within this `spillTime` counter:

https://github.com/NVIDIA/spark-rapids/blob/branch-23.10/sql-plugin/src/main/scala/com/nvidia/spark/rapids/DeviceMemoryEventHandler.scala#L120

We start the `spillTime` on line 120, and the synchronize happens on line 138, and we stop the timer on line 170. You were wondering why this time wasn't in the Spark UI metrics, and so I am curious if this is the same synchronize or something different.

**Comment by revans2** [comment]
I don't think that it is the deviceSynchronize that is causing problems. I think that this is the perfect storm in terms of memory allocation. If I disable the device synchronize the performance is much worse. If I try to use arena the performance is much worse. If I use no pooling the performance is not great, but at least it is consistent.

I got an nsys trace of a bad run and a good run (where I tuned the parallelism and batch size).  The CUDA API summary report is quite interesting.

On the bad run `cudaDeviceSynchronize` took 5.3 seconds, and the good run never called it.  But 5 seconds does not explain much about the time taken. I am going to ignore `cudaStramSynchronize_ptsz`, and instead look at the next highest time for the slow run`cudaMallocFromPoolAsync_ptsz_v11020`. On the slow run this took 702.952 seconds vs 28.798 seconds for the good run. What is more the good run actually called it more often, because the batch size is smaller. That is enough to even offset the number or retried calls. So the average time per call changed from 0.0057 ms to 0.1541 ms. That is 27x slower. The other memory related APIs are also similarly slower. I think that as the GPU fills up the allocator has to do more work to find space for an allocation, which slows the allocations down by a lot.

But I am not 100% sure on that. Because it is not just the cuda*Aync memory APIs that are slower. cudaMallocHost is 10x slower. cudaFreeHost is 3x slower.  The kernels are actually a little faster, but not massively so. I think the speedup is because less of them are running, fewer batches. 

It is almost as if interacting with the page table on the GPU from the CPU is the bottleneck. Like we are locking it when we try to do an allocation, and for large GPUs it takes forever to find a free page, and in turn slows all of the other allocations down.

**Comment by revans2** [comment]
I filed https://github.com/NVIDIA/spark-rapids/issues/9139 to look at and fix the semaphore problem properly.  I honestly don't know how to fix the performance problems here without a major change to how we deal with low memory setups. I don't know if the is something the cuda team can do, or if we need a way to detect that we are in a situation like this and try to throttle down the parallelism like we do with split and retry.

It is also looks like retry might be too aggressive with trying to have threads start to run again, because if I disable the device synchronize and just rely on retry to throttle the threads it is not doing it very well. The performance is worse than the device synchronize, and far worse than running with 3 in parallel to begin with.

Perhaps also there are metrics we could produce to detect this type of situation so the auto-tuner could also help avoid this type of situation.
